pkg/monitoring/monitoring.go

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (c) 2025-present Snowplow Analytics Ltd.
+ * All rights reserved.
+ *
+ * This software is made available by Snowplow Analytics, Ltd.,
+ * under the terms of the Snowplow Limited Use License Agreement, Version 1.1
+ * located at https://docs.snowplow.io/limited-use-license-1.1
+ * BY INSTALLING, DOWNLOADING, ACCESSING, USING OR DISTRIBUTING ANY PORTION
+ * OF THE SOFTWARE, YOU AGREE TO THE TERMS OF SUCH LICENSE AGREEMENT.
+ */
+
+package monitoring
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type MonitoringEvent struct {
+	Schema string         `json:"schema"`
+	Data   MonitoringData `json:"data"`
+}
+
+type MonitoringData struct {
+	AppName    string            `json:"appName"`
+	AppVersion string            `json:"appVersion"`
+	Tags       map[string]string `json:"tags"`
+
+	Message string `json:"message,omitempty"`
+}
+
+// MonitoringSender describes the interface for how to send heartbeat & alert events
+type MonitoringSender interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Monitoring holds a new client and supporting data for sending heartbeat & alert events
+type Monitoring struct {
+	appName           string
+	appVersion        string
+	client            MonitoringSender
+	endpoint          string
+	tags              map[string]string
+	heartbeatInterval time.Duration
+	alertChan         chan error
+	log               *logrus.Entry
+	alertInterval     time.Duration
+
+	exitSignal chan struct{}
+}
+
+func NewMonitoring(appName, appVersion string, client MonitoringSender, endpoint string, tags map[string]string, heartbeatInterval, alertInterval time.Duration, alertChan chan error) *Monitoring {
+	return &Monitoring{
+		appName:           appName,
+		appVersion:        appVersion,
+		client:            client,
+		endpoint:          endpoint,
+		tags:              tags,
+		heartbeatInterval: heartbeatInterval,
+		log:               logrus.WithFields(logrus.Fields{"name": "Monitoring"}),
+		alertChan:         alertChan,
+		alertInterval:     alertInterval,
+		exitSignal:        make(chan struct{}),
+	}
+}
+
+func (m *Monitoring) Start() {
+	var once sync.Once
+
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	ticker := time.NewTicker(m.heartbeatInterval)
+
+	go func() {
+
+		alertCooldown := time.NewTicker(time.Millisecond * 100)
+	OuterLoop:
+		for {
+			select {
+			case <-ticker.C:
+				m.log.Info("Sending heartbeat")
```

Comment:
Is there anything to stop heartbeat events from being sent if we have a setup error?